### ISSUE TITLE
dispersions examples

### ISSUE THREAD

**Issue by rugfri** [MarJMue/Berreman4x4#98]
Dear pyElli's,

Not sure this is the right place where to post this message/request. If that is the case, please excuse me.
In the Basic usage example I suspect there is a small error in the declaration of the lmfit params, the SiO2_d param is not declared.
Also, could you please provide short examples for a few more dispersion models?
I am trying to run a LorentzEnergy dispersion model (with isotropic material for now) but I keep getting an error saying " InvalidParameters: Too many parameters".

Here is the way I did building on the basic usage example

Parameters definition:
params = ParamsHist()
params.add("layer1_A", value=250, min=0, max=1000, vary=False)
params.add("layer1_E", value=57.0, min=5, max=100, vary=False)
params.add("layer1_gamma", value=5, min=1, max=7, vary=False)
params.add("layer1_t", value=230, min=200, max=260, vary=False)

Dispersion 
@fit(psi_delta_1, params)
def model(lbda, params):
    BTO = elli.LorentzEnergy(
        params["layer1_A"],
        params["layer1_E"],
        params["layer1_gamma"],
    ).get_mat()
    structure = elli.Structure(
        elli.AIR, 
        [elli.Layer(layer1, params["layer1_t"])],  
        Si,
    )
    return structure.evaluate(lbda, angle, solver=elli.Solver2x2)

Hope you can help.
Thank you for your great job!

Ruggero

**Comment by domna** [comment]
Hey Ruggero,

thank you for your report. However, I had a look at the Basic Usage example and found the SiO2_d parameter correctly declared (see [the docs](https://pyelli.readthedocs.io/en/stable/auto_examples/plot_01_basic_usage.html#sphx-glr-auto-examples-plot-01-basic-usage-py) or [ipynb example](https://github.com/PyEllips/pyElli/blob/master/examples/Basic%20Usage/Basic%20Usage.ipynb)) and the example works for me, too. Can you maybe point me to the exact example you were using?

Regarding your LorentzEnergy example:
You need to slightly change how you set the model. We distinguish between parameter you can give to a model only once and parameters you can add multiple times (as the oscillators of LorentzEnergy).
In the [general introduction](https://pyelli.readthedocs.io/en/stable/dispersions.html#dispersions) for dispersions you find more information how this is implemented.
For your example you need to change
```
BTO = elli.LorentzEnergy(
  params["layer1_A"],
  params["layer1_E"],
  params["layer1_gamma"],
).get_mat()
```

to

```
BTO = elli.LorentzEnergy().add(
  params["layer1_A"],
  params["layer1_E"],
  params["layer1_gamma"],
).get_mat()
```
and it should work.

**Comment by domna** [comment]
Hey, I just had another look at your example and found another problem. You built the layer with `elli.Layer(layer1, params["layer1_t"])`. You, however, defined the `layer1` only as a prefix for your parameters but you have to pass the dispersive material from the assignment 
```
BTO = elli.LorentzEnergy().add(
 ...
).get_mat()
```
to the `Layer`. I know it is a bit confusing since lmfit uses this prefix notation. The full example would look now like:
```
@fit(psi_delta_1, params)
def model(lbda, params):
    BTO = elli.LorentzEnergy(
        params["layer1_A"],
        params["layer1_E"],
        params["layer1_gamma"],
    ).get_mat()
    structure = elli.Structure(
        elli.AIR,
        [elli.Layer(BTO, params["layer1_t"])],
        Si,
    )
    return structure.evaluate(lbda, angle, solver=elli.Solver2x2)
```
This should work now. But please feel free to ask if you still have problems with this.

I think we will also provide a second example where we show a model with multiple input parameter sets through the `.add(...)` method.

Regarding on your concern for the right channel: Typically, these issues are used to point out bugs or feature requests for pyelli. However, since we do not have any channel to post usage questions (i.e. mailing list or a forum) it is fine to post those questions here for now, too. Eventually, we will put up another channel where people can ask such questions.

**Comment by rugfri** [comment]
Hi,

thank you.
You are right about the basic usage example, I see both sources correct now. Not sure anymore where I was reading.
About the Lorentz model, it works now. Yes, I pasted only  was incomplete.
I am now trying to reproduce the results obtained with the Woollam software on the same dataset, but couldn't so far. 
Did you do some sort of benchmarking with any existing software?
Thanks!

**Comment by domna** [comment]
Yes we did benchmark for a part of the models but mainly with spectraray, which is a software for sentech ellipsometers. Often models are slightly different and the parameters need to be adapted. We mainly use the formulas from the book Spectroscopic Ellipsometry by Fujiwara. However, you can just have a look at the formulas in the documentation and compare them to the ones Woolam provides (they are written in their manual). If I have a bit of time tomorrow I'll have a look at woolams formula and can give you guidance to adapt your params.

**Comment by domna** [comment]
Hey, I had a look at the Woollam [manual](https://physlab.lums.edu.pk/images/f/f4/Ellipsometry_manual.pdf) and I've seen that they use the formula

$$
\epsilon_\text{Lorentz} = \frac{\text{Amp}_n \text{Br}_n E_n}{E_n^2 - E^2 - i \cdot E \cdot \text{Br}_n}
$$

However, [the formula we use](https://pyelli.readthedocs.io/en/stable/dispersions.html#id1) is slightly different. Basically, the parameter $A = \text{Amp}_n * \text{Br}_n$ and `gamma = Br_n`.

Could you test to change it and see if it works then?

**Comment by rugfri** [comment]
Hi, thanks! 
Do you mean adapt the parameters or change the equation?
First attempt adapting the parameters improved significantly but the model is not yet on the data.
In pyElli/Fujiwara eq. there is also a "+1" term...not sure how this is accounted for.
Moreover I could have a look at the way the fit was done on the Wollam, there is also a e1(inf) parameter which is not described in the manual. It seems to be a freq. independent term. 
Sorry, I know all this may be confusing, nevertheless I am trying to use your code as it looks good and useful for me. Hope can make things working soon.

**Comment by domna** [comment]
Hey,

I meant to adapt the parameters. You can use the parameters from your Woollam program and just calculate the ones pyElli uses.

Yes, typically models vary in constant offset. We decided to use just `+1` and allow adding an epsilon infinity constant to the model (see [here](https://pyelli.readthedocs.io/en/stable/dispersions.html#elli.dispersions.EpsilonInf)).
You can just add dispersions together, e.g. for your example it could look like the following:
```
BTO = (elli.LorentzEnergy().add(
        params["layer1_A"],
        params["layer1_E"],
        params["layer1_gamma"],
    ) + elli.EpsilonInf(params["layer1_eps_inf"])).get_mat()
```
Here, you'd need to introduce the additional parameter `layer1_eps_inf`. Please note that it would actually evaluate to 1 + eps_inf, because Lorentz already contains the offset of 1 (sorry for the confusion but we did not want to change the model as it is used in Fujiwara and force the users to always add an `EpsilonInf(1)`).

Don't worry about your questions. It is valuable input for us, too. So we know whether we should extent and clarify the documentation somewhere or provide more examples regarding a certain topic. So please feel free to point out problems/unclear explanations in the docs as well.

**Comment by rugfri** [comment]
Thanks,

After removing the second parenthesis before .get_mat(), as it didn't match, I get

AttributeError: 'DispersionSum' object has no attribute 'get_tensor'.

**Comment by domna** [comment]
The second parenthesis is important, please not that there is an open parenthesis before `elli.LorentzEnergy` now, which you probably forgot. Here I added a few newlines for clarity:
```
BTO = (
   elli.LorentzEnergy().add(
        params["layer1_A"],
        params["layer1_E"],
        params["layer1_gamma"],
    ) 
   + elli.EpsilonInf(params["layer1_eps_inf"])
).get_mat()
```

However, adding a material and a dispersion should be an error, I will also add a proper error handling to that, so that the error message is clearer (see Issue #99)

**Comment by domna** [comment]
I just added a proper error handling for this case and found that we also have a short-hand notation for adding epsilon infinity I forgot about. You can also write:
```
BTO = (
   elli.LorentzEnergy().add(
        params["layer1_A"],
        params["layer1_E"],
        params["layer1_gamma"],
    ) 
   + params["layer1_eps_inf"]
).get_mat()
```
to add a constant offset and the `EpsilonInf` class will be automatically created for you.

**Comment by rugfri** [comment]
Thanks!! 
Things works and I will have a closer look in the next days. 
One more thing is about the usage of the refractive index database.
I have a dataset from 600 to 1700 nm, but I did not find a database entry for Silicon covering the entire range.  Is it possible to merge different entries to cover a larger spectral range? 
I know the database values may be different,  however how can handle such situation?

**Comment by domna** [comment]
Hey,

nice that it works now!

In principle it should be possible to merge different entries together, but probably requires some manual concatenating of arrays. I never really tested it but I think it is an interesting use case. I'll later prepare a small example how this could look like.

**Comment by domna** [comment]
Hey,

I created a quick example:
```
import elli

rii = elli.db.RII()
si = rii.load_dispersion("Si", "Aspnes")
si2 = rii.load_dispersion("Si", "Schinke")

si_concat = si.single_params['n'].combine_first(si2.single_params['n'])
si_disp = elli.table_index.Table(lbda=si_concat.index, n=si_concat.values)
```

From which you could use the Si dispersion as a material with `si_disp.get_mat()` or view the dispersion in a plot with `si_disp.get_dielectric_df(si_concat.index).plot()` to see if there are any jumps or other inconsistencies in the data.
Please note that accessing the `single_params` dict directly is not the best approach here. For example, if you load a dispersion which is not tabulated as nk values you would need to use `.single_params['eps']` or other fields, which depends on the particular tabulated dataset being used. I think eventually it would be nice to have support for adding tabulated datasets together, i.e. simply doing `si + si2` from the above example (which actually works but gives an error if we use the whole integration range, because internally there are still two different interpolation ranges).

If you let me know which literature sets you'd like to use I can also check the specific example for you. For the above example it actually works quite nicely without any major jumps or irregularities (there are a few fluctuations in the regions where the two dispersion overlap, though - an option could also be to not let them overlap but just use either one of the dispersions in this regions).

Edit: I added an issue for supporting adding of tabulated datasets. See #100 .

**Comment by rugfri** [comment]
Hi, many thanks!

what is "conc" in the last line?
maybe I'm again doing a mistake?

**Comment by domna** [comment]
Sorry, my bad. This is `si_concat`. I renamed it while copying it over to be more clear but forgot to rename it (I should just not do this). I edited the post above.

**Comment by rugfri** [comment]
I just ran the "basic usage" example with the si_disp and the fit does not works as good, is for the overlaps you mentioned, i gues (?)
As I have to fit a dataset up to 1700 nm, so merging the "Schinke" entry with "Salzberg and Villa" should work. The latter has only n, but k=0. 
I am not sure how to merge only specific ranges of two entries. Could you please help here?

**Comment by domna** [comment]
Yes, there you already have to approach it differently since `Salzberg and Villa` is not tabulated but is a Sellmeier function. Therefore, you can provide it a range to calculate your desired wavelength range directly. Here is a snippet how this can look like:

```python
import elli
from elli.fitting import ParamsHist, fit
import numpy as np

rii = elli.db.RII()
si_schinke = rii.load_dispersion("Si", "Schinke")
si_salzberg = rii.load_dispersion("Si", "Salzberg")


# This builds the Salzberg dispersion from the highest lambda in the Schinke tabulated data
# with 1000 points until 1700 nm. You may change the values according to your needs.
min_wavelength = si_schinke.single_params['lbda'][-1]
max_wavelength = 1700
no_points = 1000

wavelengths = np.linspace(min_wavelength, max_wavelength, no_points)
si_salzberg_table = si_salzberg.get_refractive_index(wavelengths)

# I refined this so you can select which dispersion to use when they overlap
# Here it selects the Schinke dispersion for overlapping values
si_concat = si_schinke.single_params['n'].combine(
    si_salzberg_table,
    lambda schinke, salzberg: schinke
)

elli.table_index.Table(lbda=si_concat.index, n=si_concat.values).get_dielectric_df(si_concat.index).plot()
```

Pleas note that this still is somewhat hacked together and eventually we will support these cases properly. But this may easily fail for different combinations of dispersions. But for this specific case it should work fine.

**Comment by rugfri** [comment]
Hi, thanks. 
If I plot I see that the si_concat data go up to 1470, not 1700. 

It seems to me that between your Lorentz oscillator formula in the documentation and Fujiwara's (eq. 5.9) there is a sign change at the denominator for the E's (real part).
But the source code [https://pyelli.readthedocs.io/en/stable/_modules/elli/dispersions/lorentz_energy.html#LorentzEnergy](url) is different. Compared to Woollam it has a sign change in the Im part.
Comparing with Woollam, the denominator is the samem while at numerator A_j = Amp_j Br_j E_j.
Do you agree? ( I'm bit confused : - ) )

**Comment by domna** [comment]
I see. The `combine` function is not doing what I expected it only takes the value from the second dataframe if it already exists in the first one.
I rewrote the part such that the full snippet now looks like this:
```python
import elli
from elli.fitting import ParamsHist, fit
import numpy as np
import pandas as pd

rii = elli.db.RII()
si_schinke = rii.load_dispersion("Si", "Schinke")
si_salzberg = rii.load_dispersion("Si", "Salzberg")


# This builds the Salzberg dispersion from the highest lambda in the Schinke tabulated data
# with 1000 points until 1700 nm. You may change the values according to your needs.
min_wavelength = si_schinke.single_params['lbda'][-1]
max_wavelength = 1700
no_points = 1000

wavelengths = np.linspace(min_wavelength, max_wavelength, no_points)
si_salzberg_table = si_salzberg.get_refractive_index(wavelengths)

# Concatenate the two dataframes
si_concat = pd.concat(
    [pd.DataFrame(si_schinke.single_params['n']),
    pd.DataFrame(si_salzberg_table, index=wavelengths)]
)
# Remove duplicates and keep the first occurence
si_concat = si_concat[~si_concat.index.duplicated(keep='first')]

elli.table_index.Table(lbda=si_concat.index, n=si_concat.values[:,0]).get_dielectric_df(si_concat.index).plot()
```

Yes, there may be a sign change, because this depends on the convention being used. You can either write your dispersions in the `n + ik` or `n - ik` convention. Woollam probably uses the other convention as we do.

Yes, I see that there is some wrong sign in the documentation. However, I had a look at the code and it is the other way around there. This is the actual code being used:
```python
c.get("A") / (c.get("E") ** 2 - energy**2 - 1j * c.get("gamma") * energy)
```

Thank you for pointing it out. I'll correct the documentation (actually we're still in the phase of checking the dispersions and their dispersions for errors - see #55 ).

**Comment by rugfri** [comment]
thanks. The interpolation give still error. It seems that the number of points in along x and y is different.
ValueError: x and y arrays must be equal in length along interpolation axis.

Another question on the dispersion models, What would be a convenient way to include an exponential tail (Urbach tail) into a Cauchy model?

**Comment by domna** [comment]
Ah yes, you need to change the last line to:
```
elli.table_index.Table(lbda=si_concat.index, n=si_concat.values[:,0]).get_dielectric_df(si_concat.index).plot()
```
to use `si_concat.values[:,0]` instead of `si_concat.values`. I corrected it in the above example, too.

I'm not sure if it is the best approach to include an Urbach tail into a Cauchy model, since this already models the part far away from any resonance and does not include a parameter for the actual bandgap of the material. Maybe it is better to directly use a Cody-Lorentz function which already includes an Urbach tail. If you really want to model a Cauchy + exponential you could create an exponential dispersion model similar to the [Poles](https://pyelli.readthedocs.io/en/stable/dispersions.html#elli.dispersions.Poles) which we already provide. I could also assist you in creating such a model.

**Comment by rugfri** [comment]
Thanks. it works now.

I see your point, thanks.
The reason is that I am trying to benchmark and the results obtained with the Woollam software for an oxide film on Silicon. The model includes also the SiO2 on top of Si (so that air, oxide, sio2, si) where the SiO2 is modelled with Cauchy and Urbach tail. Was thinking to model correction/improvement as a second step. 
So, I think I will try create an "Urbach tail" class (following the Poles example).  Thanks for the offer of support, I will need it : - )

**Comment by domna** [comment]
Hey,

I had a look at the Woollam manual. You are referring to the Cauchy_Extended model, right? The problem is that they don't provide the formula for this model. If you can share your data + theoretical curve + model parameters I could have a look and see if I could reverse engineer this model. From their manual it seems to be a combination of an IR pole and a Cody part accounting for the Urbach tail (similar to the Cody-Lorentz model but just using a Cauchy to model the resonance instead of a Lorentz oscillator).

Would it generally of interest for you if we provided the Woollam models as well? For example in some submodule like `elli.dispersions.woollam` or so.

**Comment by rugfri** [comment]
I refer to the Cauchy model, eq. 8-1 and 8-2 (p. 9-288).
Yes, in general  it could be useful having a Woollam submodule. 
Right now I aim only at reproducing a result, and that Urbach tail is the most significant difference I see (beside the slight difference in the definition of the Lorentz model). I would proceed step-wise, unless for you is easy and fast to implement the Wollam submodule.

**Comment by domna** [comment]
Hey, no unfortunately we would also need to implement the models by hand. I was rather asking if it would be beneficial for you in the long term to have a special module from which you can access the exact same function as Woollam provides.

It would also be nice if you would contribute your model as soon as you get it working.

**Comment by rugfri** [comment]
Would implementing a "cauchy_urbach" dispersion with n, k given by Fujiwara's eq. 7.6a, 7.6b a possibility?

**Comment by MarJMue** [comment]
Hi Ruggero, 
I appreciate all of your feedback and am very interested how pyelli compares with the woolam software.
I can add the CauchyUrbach dispersion.

Marius

**Comment by MarJMue** [comment]
This should work for now. I will include it in the package, but writing tests and documentation takes some more time.

    import elli
    import numpy as np

    class CauchyUrbach(elli.Dispersion):
        r"""Cauchy dispersion, with an Urbach Tail absorption.

        Single parameters:
            :n0: Defaults to 1.5.
            :B: Defaults to 0. Unit in 1/eV\ :sup:`2`.
            :C: Defaults to 0. Unit in 1/eV\ :sup:`4`.
            :D: Defaults to 0.
            :Eg: Defaults to 0. Unit in eV.
            :Eu: Defaults to 1. Unit in eV.
    
        Repeated parameters:
            --

        Output:
        """

        single_params_template = {"n0": 1.5, "B": 0, "C": 0, "D": 0, "Eg": 0, "Eu": 1}
        rep_params_template = {}

        def dielectric_function(self, lbda):
            energy = elli.utils.conversion_wavelength_energy(lbda)
            refr_index = (
                self.single_params.get("n0")
                + self.single_params.get("B") * energy**2
                + self.single_params.get("C") * energy**4
                + 1j
                * self.single_params.get("D")
                * np.exp(
                    (energy - self.single_params.get("Eg")) / self.single_params.get("Eu")
                )
            )
            return refr_index**2

**Comment by domna** [comment]
Hey @rugfri, @MarJMue ,

I had a look at Fuijiwara, too. There is a slightly different formula in `Spectroscopic Ellipsometry for Photovoltaics: Vol 1` (5.15 & 5.16):
$$n(\lambda) = A + \frac{B}{\lambda^2} + \frac{C}{\lambda^4} - D \lambda^2$$
and
$$k(E) = \alpha \exp(\beta (E - E_g))$$

It looks like the model Woollam is using as it also includes the NIR term `-D\lambda^2`. So maybe we should try this for your model, too @rugfri?

**Comment by rugfri** [comment]
Hi both and thanks!
@domna, yes, I will try that too. Though from what I see the results I have were derived with the eqs. 8-1 and 8-2 (p. 9-288), where  only k(E) is only slightly different from the one @MarJMue implemented but should give consistent behaviour. 

I guess now I should just add this dispersion to my local copy and test, correct? There is not yet update.

Slightly off-topic, I cloned locally pyElli, but the installation of the package in dev mode returned 
"no matches found: .[fitting,testing]".

thanks.

**Comment by MarJMue** [comment]
> I guess now I should just add this dispersion to my local copy and test, correct? There is not yet update.

You can just run the code snipped before your code in your scipt/jupyter notebook

**Comment by domna** [comment]
> Slightly off-topic, I cloned locally pyElli, but the installation of the package in dev mode returned "no matches found: .[fitting,testing]".

I think in some shells you have to put quotes around it like so `pip install -e ".[fitting,testing]"`. Maybe we should just write it with quotes in the docs

**Comment by domna** [comment]
@rugfri I'll close this issue now as your issues seem to be resolved. Feel free to re-open this or another issue if you need further help.

**Comment by rugfri** [comment]
Hi, had been detached from ellipsometry for a while but now I am back. 
I retrieved some test SiO2 data for the Woollam - pyElli bench-marking that I haven't been able to complete yet.
With the Wollam software the data get fitted with a Cauchy model.
With pyElli I can get close to Psi but not to Delta.
Would you be available to help me on this? I can share the experimental data here if you wish, in case.
Thanks!

**Comment by domna** [comment]
Hey @rugfri,

yes, sure I'll assist you. Indeed, it would be easiest if you'd share your measurement data with me. Probably, it is just another representation of delta from the Woollam software (it may be in -180 to 180 or 0 to 360 deg range or can have a +i or -i convention - but I can figure it out from your data).
We recently also decided to open a discord channel to address questions there, so please feel free to also join there (see https://github.com/PyEllips/pyElli#got-a-question).
